akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Blueprint.scala

```diff
@@ -29,60 +28,50 @@ import scala.collection.immutable
 /**
  * Represents one direction of an Http2 substream.
  */
-private[http2] sealed trait Http2SubStream {
-  type T
-  val initialHeaders: ParsedHeadersFrame
-  val data: Source[T, Any]
-  val correlationAttributes: Map[AttributeKey[_], _]
+@InternalApi
+private[http2] case class Http2SubStream(
+  initialHeaders:        ParsedHeadersFrame,
+  data:                  Source[Any, Any],
+  correlationAttributes: Map[AttributeKey[_], _]
+) {
   def streamId: Int = initialHeaders.streamId
-
-  def withCorrelationAttributes(attributes: Map[AttributeKey[_], _]): Http2SubStream
-
-  def createRequestEntity(contentLength: Long, contentType: ContentType): RequestEntity =
-    this match {
-      case s if s.data == Source.empty || contentLength == 0 => HttpEntity.Empty
-      case s: ByteHttp2SubStream =>
-        if (contentLength > 0) HttpEntity.Default(contentType, contentLength, s.data)
-        else HttpEntity.Chunked.fromData(contentType, s.data)
-      /* contentLength undefined */
-      case c: ChunkedHttp2SubStream => HttpEntity.Chunked(contentType, c.data)
+  def hasEntity: Boolean = !initialHeaders.endStream
+
+  def withCorrelationAttributes(newAttributes: Map[AttributeKey[_], _]): Http2SubStream =
+    copy(correlationAttributes = newAttributes)
+
+  def createEntity(contentLength: Long, contentTypeOption: OptionVal[ContentType]): RequestEntity = {
+    def contentType: ContentType = contentTypeOption.getOrElse(ContentTypes.`application/octet-stream`)
```

Comment:
Why this is by default set to application/octet-stream? Some client like K9s do not set Content-type header and on the k8s server side it defaults missing content type header to application/json. Using a code written with akka-http to proxy this connection leads to errors.

Reply:
The content type is not optional in Akka HTTP, so we have to choose something. I suspect this is consistent with the HTTP/1.1 implementation, but I haven't checked - do you know?

Reply:
Yes, it's the same for our HTTP/1.1 implementation: 

https://github.com/akka/akka-http/blob/eb7bc5803ac1ed561f85fff961abc3203ae3cefc/akka-http-core/src/main/scala/akka/http/impl/engine/parsing/HttpMessageParser.scala#L357-L360

[RFC 7231](https://datatracker.ietf.org/doc/html/rfc7231#section-3.1.1.5) says:

> If a Content-Type header field is not present, the recipient
> MAY either assume a media type of "application/octet-stream"
> ([RFC2046], Section 4.5.1) or examine the data to determine its type.